### ISSUE TITLE
souce code question!

### ISSUE THREAD

**Issue by fogin** [moonlight-stream/moonlight-qt#1494]
// Transfer ownership into ComPtrs
    for (int i = 0; i < m_DecSurfaces.size(); i++) {
        m_DecSurfaces[i].Attach(m_DecSurfacesRaw[i]);
    }

    hr = m_DecService->CreateVideoDecoder(chosenDeviceGuid, &m_Desc, &m_Config,
                                          m_DecSurfacesRaw.data(),
                                          (UINT)m_DecSurfacesRaw.size(),
                                          &m_Decoder);

is m_DecSurfacesRaw available after Transfer ownership into m_DecSurfaces?

**Comment by cgutman** [comment]
Yeah, it is. The semantics of `ComPtr::Attach()` are that the lifetime ownership is transferred to the `ComPtr` but the original pointers remain valid as long as the `ComPtr` hasn't free them (which it won't there when we use them). IOW, you can think of it as a move semantic for the ownership of the object, but just a copy semantic for the pointer itself.
